crates/exex/exex/src/dyexex.rs

```diff
@@ -0,0 +1,167 @@
+        > {
+            Box::pin($user_fn(ctx))
+        }
+    };
+}
+
+#[derive(Debug, Default)]
+/// Dynamic ExEx loader
+pub struct DyExExLoader<Fut> {
+    /// List of loaded Dynamically Loaded ExExes
+    pub loaded: Vec<LoadedExEx<Fut>>,
+}
+
+impl<Fut> DyExExLoader<Fut>
+where
+    Fut: Future<Output = Result<()>> + Send,
+{
+    /// Initializes a Dynamic ExEx loader.
+    /// TODO(0xurb) - path to datadir here and loaded list with capacity of dir inner files.
+    pub const fn new() -> Self {
+        Self { loaded: Vec::new() }
+    }
+
+    /// Loads a dynamically loaded ExEx from a given path.
+    ///
+    /// # Safety
+    ///
+    /// The dynamically loaded ExEx library **must** contain a function named `_launch_exex`
+    /// with the correct type signature, that resolves to exact function pointer.
+    ///
+    /// Otherwise, behavior is undefined.
+    /// See also [`Library::get`] for more information on what
+    /// restrictions apply to `_launch_exex` symbol.
+    #[allow(clippy::type_complexity)]
+    pub unsafe fn load(&mut self, path: impl AsRef<Path>, ctx: ExExContextDyn) -> Result<()> {
+        let path = path.as_ref();
+
+        // Dynamically loaded ExEx id it's a filename with stripped [`DLL_PREFIX`] and
+        // [`DLL_SUFFIX`]
+        let lib_name = path
+            .file_name()
+            .ok_or_eyre("cannot obtain a filename for dyexex")?
+            .to_str()
+            .ok_or_eyre("dyexex name is not a valid UTF-8 encoded string")?
+            .strip_prefix(DLL_PREFIX)
+            .ok_or_else(|| {
+                eyre::eyre!(
+                    "dyexex is not a shared dynamic library (has no prefix: `{DLL_PREFIX}`)"
+                )
+            })?
+            .strip_suffix(DLL_SUFFIX)
+            .ok_or_else(|| {
+                eyre::eyre!(
+                    "dyexex is not a shared dynamic library (has no suffix: `{DLL_SUFFIX}`)"
+                )
+            })?;
+
+        let lib = Library::new(path)?;
+        let symbol: Symbol<
+            '_,
+            unsafe fn(ExExContextDyn) -> *mut (dyn Future<Output = Fut> + Send),
+        > = lib.get(LAUNCH_EXEX_FN)?;
+
+        let raw_func_pointer = symbol(ctx);
+        if raw_func_pointer.is_null() {
+            return Err(eyre::eyre!("Failed to load future from dynamic library"));
+        }
+
+        // SAFETY: We guarantee that the pointed data is pinned, loaded DLL symbol
+        // function pointer resolves to the pinned future.
+        let exex_fut = Pin::new_unchecked(Box::from_raw(raw_func_pointer));
+
+        self.loaded.push(LoadedExEx::new(lib_name, lib, exex_fut));
+
+        Ok(())
+    }
+}
+
+/// Dynamically Loaded ExEx representation
+pub struct LoadedExEx<Fut> {
+    /// ExEx id
+    id: String,
+    /// Loaded [`Library`] pointer
+    lib: Arc<Library>,
+    /// Regarding ExEx launch entrypoint future
+    exex_fut: Pin<Box<dyn Future<Output = Fut> + Send>>,
```

Comment:
hmm, can we do it with a generic here? Let's just box the inner future as well, and make the `define_exex` macro pin it?